### PR TITLE
Add bit shift operators and enable bitwise not

### PR DESCRIPTION
**Pull request by guipublic**
I added support for bit shift left and right operators into the frontend and enabled prefix expression for bitwise not.

### CODE REVIEW

File: crates/noirc_frontend/src/ast/expression.rs [hunk 1]
```diff
@@ -244,6 +246,8 @@ impl BinaryOpKind {
             BinaryOpKind::Or => "|",
             BinaryOpKind::Xor => "^",
             BinaryOpKind::Assign => "=",
+            BinaryOpKind::Shr => ">>",
+            BinaryOpKind::Shl => "<<",
```
Comment (jfecher) [inline crates/noirc_frontend/src/ast/expression.rs:250]: Can we expand the name of these operators to `ShiftLeft` and `ShiftRight`, the other op kinds aren't abbreviated

File: crates/noirc_frontend/src/parser/mod.rs [hunk 2]
```diff
@@ -248,6 +248,8 @@ impl Precedence {
             Token::Pipe => Precedence::Sum,
             Token::Plus => Precedence::Sum,
             Token::Minus => Precedence::Sum,
+            Token::ShiftLeft => Precedence::Sum, //TODO checks rust precedence which seems to be: sum >shift>and>xor>or
+            Token::ShiftRight => Precedence::Sum,
```
Comment (jfecher) [inline crates/noirc_frontend/src/parser/mod.rs:252]: These seem correct, can we add these new precedence levels now?

File: crates/noirc_evaluator/src/lib.rs [hunk 3]
```diff
@@ -169,6 +169,10 @@ impl<'a> Evaluator<'a> {
                 let err = RuntimeErrorKind::Unimplemented("The Or operation is currently not implemented. First implement in Barretenberg.".to_owned());
                 Err(err)
             }
+            HirBinaryOpKind::Shr | HirBinaryOpKind::Shl => {
+                let err = RuntimeErrorKind::Unimplemented("Bit shift operations are not currently implemented.".to_owned());
+                Err(err)
```
Comment (jfecher) [inline crates/noirc_evaluator/src/lib.rs:174]: nitpick but lets get rid of the `err` variable here, it's name is unhelpful and it'd be shorter to just wrap the RuntimeErrorKind in Err.

File: crates/noirc_frontend/src/hir/type_check/expr.rs [hunk 4]
```diff
@@ -226,9 +226,8 @@ pub(crate) fn type_check_expression(
 
             block_type
         }
-        HirExpression::Prefix(_) => {
-            // type_of(prefix_expr) == type_of(rhs_expression)
-            todo!("prefix expressions have not been implemented yet")
+        HirExpression::Prefix(prefix_expr) => {
+            type_check_expression(interner, &prefix_expr.rhs, errors)
```
Comment (jfecher) [inline crates/noirc_frontend/src/hir/type_check/expr.rs:230]: This is the only change I can find that would relate to a bitwise not. What is the plan for `!`? We want it now to work for booleans and as a bitwise operator like we do for `&` and `|`?

While we're fixing this code can we also get proper type checking for prefix operators? We should have a new function which checks if the rhs is numeric for `-` or if it is numeric or a boolean for `!`. Both should return the same type as their operand as well.

File: crates/noirc_evaluator/src/lib.rs (hunk 3 above)
Comment (guipublic) [reply crates/noirc_evaluator/src/lib.rs:174]: Done

File: crates/noirc_frontend/src/ast/expression.rs (hunk 1 above)
Comment (guipublic) [reply crates/noirc_frontend/src/ast/expression.rs:250]: Done

File: crates/noirc_frontend/src/hir/type_check/expr.rs (hunk 4 above)
Comment (guipublic) [reply crates/noirc_frontend/src/hir/type_check/expr.rs:230]: Yes, `!` will be the bitwise not

File: crates/noirc_frontend/src/hir/type_check/expr.rs [hunk 5]
```diff
@@ -348,6 +359,22 @@ fn type_check_function_call(
     }
 }
 
+pub fn prefix_operand_type_rules(op: &HirUnaryOp, rhs_type: &Type) -> Result<Type, String> {
+    match op {
+        HirUnaryOp::Minus => {
+            if !matches!(rhs_type, Type::Integer(..)) {
```
[the remark this answers is not in the document]
Comment (guipublic) [reply crates/noirc_frontend/src/hir/type_check/expr.rs]: Done

**Review by jfecher — APPROVED** [review]

**Pull request merged**